### PR TITLE
Add metrics server as addon to k8s CI

### LINKED ISSUE #1125: Add Metrics server to clusters
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
After deploying a cluster the commands `kubectl top nodes` and `kubectl top pods` don't work because the [metric server](https://github.com/kubernetes-sigs/metrics-server) is missing.  Functionality like the [HPA will also require](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/#before-you-begin) the metric server.  

```
 k top nodes
error: Metrics API not available
 k top pods
error: Metrics API not available
 k get pods -A
NAMESPACE     NAME                                                           READY   STATUS    RESTARTS   AGE
kube-system   calico-kube-controllers-8f59968d4-v8clz                        1/1     Running   0          60m
kube-system   calico-node-clfb2                                              1/1     Running   0          60m
kube-system   calico-node-kdv67                                              1/1     Running   0          58m
kube-system   calico-node-n542l                                              1/1     Running   0          57m
kube-system   coredns-f9fd979d6-gf6hr                                        1/1     Running   0          60m
kube-system   coredns-f9fd979d6-h2m7m                                        1/1     Running   0          60m
kube-system   etcd-default-template-control-plane-9przl                      1/1     Running   0          60m
kube-system   kube-apiserver-default-template-control-plane-9przl            1/1     Running   0          60m
kube-system   kube-controller-manager-default-template-control-plane-9przl   1/1     Running   0          60m
kube-system   kube-proxy-4pb22                                               1/1     Running   0          60m
kube-system   kube-proxy-qf5v7                                               1/1     Running   0          57m
kube-system   kube-proxy-xqrvw                                               1/1     Running   0          58m
kube-system   kube-scheduler-default-template-control-plane-9przl            1/1     Running   0          60m
```

```
 kubectl describe hpa
Name:                                                  php-apache
Namespace:                                             default
Labels:                                                <none>
Annotations:                                           <none>
CreationTimestamp:                                     Fri, 15 Jan 2021 13:17:36 -0800
Reference:                                             Deployment/php-apache
Metrics:                                               ( current / target )
  resource cpu on pods  (as a percentage of request):  <unknown> / 50%
Min replicas:                                          1
Max replicas:                                          10
Deployment pods:                                       1 current / 0 desired
Conditions:
  Type           Status  Reason                   Message
  ----           ------  ------                   -------
  AbleToScale    True    SucceededGetScale        the HPA controller was able to get the target's current scale
  ScalingActive  False   FailedGetResourceMetric  the HPA was unable to compute the replica count: unable to get metrics for resource cpu: unable to fetch metrics from resource metrics API: the server could not find the requested resource (get pods.metrics.k8s.io)
Events:
  Type     Reason                        Age                From                       Message
  ----     ------                        ----               ----                       -------
  Warning  FailedGetResourceMetric       14s (x4 over 60s)  horizontal-pod-autoscaler  unable to get metrics for resource cpu: unable to fetch metrics from resource metrics API: the server could not find the requested resource (get pods.metrics.k8s.io)
  Warning  FailedComputeMetricsReplicas  14s (x4 over 60s)  horizontal-pod-autoscaler  invalid metrics (1 invalid out of 1), first error is: failed to get cpu utilization: unable to get metrics for resource cpu: unable to fetch metrics from resource metrics API: the server could not find the requested resource (get pods.metrics.k8s.io)

```

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

This could be applied via [Custom Resource Sets](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20200220-cluster-resource-set.md) similar to the way we do CNI's: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e7ebdf3bef11fcb5a6e001b2860ae8d6654e0202/Tiltfile#L199

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/templates/addons/calico-resource-set.yaml

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by jsturtevant**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
We would like to run HPA tests in upstream and the metrics server is required for HPA components: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1881

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1125

**Special notes for your reviewer**:
I've only added this to the CI Job that is need for now but did tried to do it in a way that would make it pretty easy to add to other jobs and tilt if we wanted.

Also using kustomize to pull in the metric server from the releases page on https://github.com/kubernetes-sigs/metrics-server/releases.  This allows for easy updates and overriding settings if needed via kustomize. This could be a pattern we use for the other addons as well.

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Add metrics server to CI jobs for upstream testing
```

### CODE REVIEW

**Comment by jsturtevant** [comment]
/hold

I've got it deploying but running into errors:

```
I1203 00:40:24.651567       1 server.go:188] "Failed probe" probe="metric-storage-ready" err="not metrics to serve"
E1203 00:40:29.522716       1 scraper.go:139] "Failed to scrape node" err="Get \"https://10.1.0.5:10250/stats/summary?only_cpu_and_memory=true\": x509: cannot validate certificate for 10.1.0.5 because it doesn't contain any IP SANs" node="capz-conf-bvvs4"
E1203 00:40:29.526596       1 scraper.go:139] "Failed to scrape node" err="Get \"https://10.1.0.4:10250/stats/summary?only_cpu_and_memory=true\": x509: cannot validate certificate for 10.1.0.4 because it doesn't contain any IP SANs" node="capz-conf-gtfr6"
E1203 00:40:29.527545       1 scraper.go:139] "Failed to scrape node" err="Get \"https://10.0.0.4:10250/stats/summary?only_cpu_and_memory=true\": x509: cannot validate certificate for 10.0.0.4 because it doesn't contain any IP SANs" node="capz-conf-5ghwmu-control-plane-9rlx5"
```

Apparently kubeadm doesn't set this up properly out of the box: https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/troubleshooting-kubeadm/#cannot-use-the-metrics-server-securely-in-a-kubeadm-cluster

It looks like we are missing the ability to properly get signed certs:
https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#kubelet-serving-certs

> apiVersion: kubeadm.k8s.io/v1beta3
kind: ClusterConfiguration
> ---
> apiVersion: kubelet.config.k8s.io/v1beta1
> kind: KubeletConfiguration
> serverTLSBootstrap: true

**Comment by jsturtevant** [comment]
looks like there has been some thought on how this works with capi: https://github.com/kubernetes-sigs/cluster-api/blob/58905fb5140a7844c3e24aa70d255fb8f9d0c491/docs/proposals/20210222-kubelet-authentication.md#changes-to-cluster-api-bootstrap-provider-kubeadm

**Comment by jsturtevant** [comment]
After following the directions in https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#kubelet-serving-certs (note that `ServerTLSBootstrap` setting in kubelet config is actually the flag `rotate-server-certificates` in [kubelet command line parameters](https://github.com/kubernetes/kubernetes/blob/a11453efbc4a5575f7945af1c6fd4f7c00379529/cmd/kubelet/app/options/options.go#L434)

setting the field `rotate-server-certificates` on the kubelet join configurations for all the nodes did work but required manual intervention to approve the CSR before metric server would work.  The cluster and the deployment still came up but with the same error as [above](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1903#issuecomment-985119956) before the manual approval.

As there is a [proposal to address this long term in CAPI](https://github.com/kubernetes-sigs/cluster-api/blob/58905fb5140a7844c3e24aa70d255fb8f9d0c491/docs/proposals/20210222-kubelet-authentication.md#changes-to-cluster-api-bootstrap-provider-kubeadm) I will plan to turn on the metric server flag `--kubelet-insecure-tls` for the e2e tests with the plan to enable this customers longer term with the CAPI solution.  It is possible to write a controller that would auto enable the CSR signing but would be not relevant once the CAPI work is in place.

**Comment by jsturtevant** [comment]
The windows tests ran but have the flakes I am investigating in s separate workstream.

/test pull-cluster-api-provider-azure-windows-containerd-upstream-with-ci-artifacts

File: templates/addons/metrics-server/kustomization.yaml
Comment (CecileRobertMichon) [inline templates/addons/metrics-server/kustomization.yaml:5]: this is awesome, we should use this model for more stuff (eg. cloud-provider) instead of maintaining addon specs in the repo

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm

**Review by devigned — APPROVED** [review]
/lgtm

+1 to how you are pulling the metrics server resources

File: templates/addons/metrics-server/kustomization.yaml
Comment (jsturtevant) [reply templates/addons/metrics-server/kustomization.yaml:5]: cool!  I created a tracking issue with a few possible items: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1917

**Pull request merged**
